### PR TITLE
Priority Queue + 2D Grid and pathfinding

### PR DESCRIPTION
**Pull request by kefniark**
## Description

* Add a `Priority Queue` (@Yuuki77 if you can have a look at this one 😄)
* Add a `SquareGrid` (class to represent a 2d grid)
* Add a simple pathfinding implementation

image

## TODO
I also plan to add:
  * ~option to have diagonal move on this SquareGrid~
  * new type of 2d grid like hexagonal (6 neighbors)
  * 3d grid (voxel)

### CODE REVIEW

File: src/algorithm/priorityQueue.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+	/**
+	 * Get the next element in the queue (first or last based on order)
+	 *
+	 * @returns {T}
+	 */
+	next() {
+		if (!this.tree.isEmpty()) return undefined
+		const entry = this.order === PriorityQueueOrder.Higher ? this.tree.max : this.tree.min
+		const res = entry.val.pop()
```
Comment (kefniark) [inline src/algorithm/priorityQueue.ts]: hum I have a doubt now, when multiple element have the same priority, it should probably be used like a queue (it's a priority queue after all) and the first added should be the first processed 🤔 
So `entry.val.shift()`

**Comment by Yuuki77** [comment]
Yeah I will take a look in this weekends!

File: src/algorithm/priorityQueue.ts (hunk 1 above)
Comment (Yuuki77) [reply src/algorithm/priorityQueue.ts]: but I think current BST implementation doesn't support the same keys..

Comment (kefniark) [reply src/algorithm/priorityQueue.ts]: Yes i know, that's why I worked around by using `T[]`  and not not `T` in this tree
So for each priority there is a unique array, but the content of this array can change and is 0..n

Comment (kefniark) [reply src/algorithm/priorityQueue.ts]: Not related to this PR but more to our binary tree

http://user.it.uu.se/~arnea/ps/simp.pdf

**Comment by kefniark** [comment]
Just added:
* And option to allow diagonal move on `SquareGrid`
* An extension of `SquareGrid` with walls and corner between tiles: `SquareGridWall`

Based on a discussion got few month ago http://disq.us/p/201er0h

This allow to create wall between adjacent tiles or columns
image

The API is really similar to the one of `SquareGrid` and pathfinding work the same

Pathfinding with diagonal moves and thin wall
image

**Comment by kefniark** [comment]
The current implementation (bfs with priority queue) works well and is quite efficient, but the path found is not always optimum
image

**Comment by kefniark** [comment]
Here we go, after just a small tweak on the heuristic, now it find the best path 👍
```ts
const heuristic = this.distanceToGoal(...)
// replaced by
const heuristic = distanceToOrigin + this.distanceToGoal(...)
```
image

**Comment by kefniark** [comment]
OK a bit of cleanup:
* get rid of LGTM alerts
* code coverage

And added:
* new predefined color palette + color
* new eventBus class

**Pull request merged**